packages/indexer/src/pinner.ts

```diff
@@ -0,0 +1,99 @@
+import { Job, Queue, Worker } from 'bullmq';
+import { parseIpfsCid, parseIpfsUrl, readRawIpfs, uncompress } from '@usecannon/builder';
+import { getS3Client } from '@usecannon/repo/src/s3';
+import { parseRedisUrl } from './helpers/parse-redis-url';
+import { config } from './config';
+
+const QUEUE_NAME = 'pinner-queue';
+
+const s3 = getS3Client(config);
+
+const processors = {
+  async PIN_CID(data: { cid: string }) {
+    // eslint-disable-next-line no-console
+    console.log('PIN_CID: ', data.cid);
+
+    const cid = parseIpfsUrl(data.cid) || parseIpfsCid(data.cid);
+    if (!cid) throw new Error(`Invalid CID ${data.cid}`);
+
+    const exists = await s3.objectExists(cid);
+    if (exists) return;
+
+    const rawData = await readRawIpfs({
+      ipfsUrl: config.IPFS_URL,
+      cid,
+    });
+
+    await s3.putObject(cid, rawData);
+  },
+
+  async PIN_PACKAGE(data: { cid: string }) {
```

Comment:
this segment of code is not sufficient to ensure that a package is published. You must also pin any subpackages (ie, `pull`ed or `clone`d package urls must also be pinned, and they may not be pinned automatically!)